### DISCUSSION TITLE
Freeze  individual weights

### DISCUSSION THREAD

**Discussion by lbelf** [espnet/espnet#4293] [Q&A]
Hi,

I want to re-estimate some parameters of a pruned model. Is there a way to freeze only some weights of a given layer?

Thank You,

**Comment by sw005320** [comment]
These pages may help
https://espnet.github.io/espnet/espnet2_training_option.html#transfer-learning-fine-tuning-using-pretrained-model
https://espnet.github.io/espnet/espnet2_training_option.html#freeze-parameters

**Reply by lbelf** [reply]
Thank you for your answer. I understood that I can use the --freeze_param option of Espnet2 to specify the individual weights to be frozen. 
The problem is that my models are trained and pruned with Espnet1 (model-module: "espnet.nets.pytorch_backend.e2e_asr_transformer:E2E"). Can I use them in Espnet2? If not, what do you suggest?

Thank you.

**Reply by sw005320** [reply]
I see.
As far as I remember, espnet1 does not have such a useful function.
Yes, maybe, it would be a good occasion to switch to espnet2.

**Reply by b-flo** [reply]
Hi,

If you want to keep using ESPnet1, we have equivalent functions for finetuning, please see https://espnet.github.io/espnet/tutorial.html#how-to-use-finetuning.
Otherwise, you will need to switch to ESPnet2 as @sw005320 said.

**Comment by lbelf** [comment]
Hi,

Yes, I prefer to continue using espnet1 for the moment (I'm just getting used to it and I wrote some code inside ...). 
On the other hand, I think it is possible to modify the "freeze_modules" module (in espnet/espnet/asr/pytorch_backend/asr_init.py) to add the ability to freeze individual weights in espnet1. For this, can you provide me with the name (or link or code) of the equivalent module in espnet2 (that is invoked via the --freeze_param option) to use?

**Comment by b-flo** [comment]
> On the other hand, I think it is possible to modify the "freeze_modules" module (in espnet/espnet/asr/pytorch_backend/asr_init.py) to add the ability to freeze individual weights in espnet1.

Correct me If I'm wrong but freezing functions in ESPnet1 and ESPnet2 are the same in terms of capabilities?

[ESPnet1](https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/asr_init.py#L30-L37=) vs. [ESPnet2](https://github.com/espnet/espnet/blob/7a7d911d900773cb7d9ee1b26ebe0bc848d600ef/espnet2/tasks/abs_task.py#L1130-L1134=)

```python
    for mod, param in model.named_parameters():
        if any(mod.startswith(m) for m in modules):
            logging.warning(f"Freezing {mod}. It will not be updated during training.")
            param.requires_grad = False

    model_params = filter(lambda x: x.requires_grad, model.parameters())
```

```python
        for t in args.freeze_param:
            for k, p in model.named_parameters():
                if k.startswith(t + ".") or k == t:
                    logging.info(f"Setting {k}.requires_grad = False")
                    p.requires_grad = False
```

The only difference is that we filter parameters with `requires_grad=False` in ESPnet1 because some old versions of torch won't freeze parameters if the parameter exists.

**Reply by lbelf** [reply]
In espnet1, the "freeze_modules" module freezes an entire layer and not just some of the weights in that layer (as I would like). I don't know if this last option is possible in espnet2. 
By the way, it seems to me that the parameter t in your espnet2 code above refers to a module and not to a weight. Doesn't it?

**Reply by b-flo** [reply]
>In espnet1, the "freeze_modules" module freezes an entire layer and not just some of the weights in that layer (as I would like). I don't know if this last option is possible in espnet2.

As far as I'm aware you can't apply `requires_grad = False` to a slice of a tensor (i.e.: weight from the module). You could zero out the parts you don't want but the weights will be trained/changed depending on the optimizer anyway.

If what you mean by "some of the weights" is "some attributes of the layer" (e.g.: `bias`, `weight_ih`, `weight_hh`, etc), you can specify them in both versions through `freeze_modules` and `freeze_param`!

**Reply by lbelf** [reply]
Hi,

and thank you for these information. Yes I want to freeze a slice of a tensor. 
I found some suggestions (in https://discuss.pytorch.org/t/freezing-part-of-the-layer-weights/9457). The proposal is to manually zero out gradient before I call optimizer function.
To this end, the following code is provided (in https://discuss.pytorch.org/t/freezing-individual-weights/57027/4)
EPS = 1e-6       
for name, p in model.named_parameters():
            if 'weight' in name:
                tensor = p.data.cpu().numpy()
                grad_tensor = p.grad.data.cpu().numpy()
                grad_tensor = np.where(tensor < EPS, 0, grad_tensor)     
                p.grad.data = torch.from_numpy(grad_tensor).to(device)

I don't know where this code can be integrated into the train()  function of the module espnet/espnet/asr/pytorch_backend/asr.py
Any help please?

**Reply by b-flo** [reply]
> I don't know where this code can be integrated into the train() function of the module espnet/espnet/asr/pytorch_backend/asr.py

In `update_core(...)` from `espnet/espnet/asr/pytorch_backend/asr.py`, after `loss.backward()` and before `optimizer.step()` as suggested in your second link I suppose?
Also, again, note that zeroing the parts you don't want is not the same as freezing as the weights will be modified  anyway if grad is required.

**Comment by lbelf** [comment]
According to this code, it is rather the gradient that is set to zero. So weights set to zeros with zero gradients will not be changed, right?
